services/invitations/pkg/backends/keycloak/backend.go

```diff
@@ -0,0 +1,149 @@
+		logger: log.Logger{
+			logger.With().
+				Str("invitationBackend", "keycloak").
+				Str("baseURL", baseURL).
+				Str("clientID", clientID).
+				Str("clientRealm", clientRealm).
+				Str("userRealm", userRealm).
+				Logger(),
+		},
+		client:       client,
+		clientID:     clientID,
+		clientSecret: clientSecret,
+		clientRealm:  clientRealm,
+		userRealm:    userRealm,
+	}
+}
+
+// CreateUser creates a user in the keycloak backend.
+func (b Backend) CreateUser(ctx context.Context, invitation *invitations.Invitation) (string, error) {
+	token, err := b.getToken(ctx)
+	if err != nil {
+		return "", err
+	}
+	u := uuid.New()
+
+	firstName, lastName := splitDisplayName(invitation.InvitedUserDisplayName)
+	b.logger.Info().
+		Str(idAttr, u.String()).
+		Str("email", invitation.InvitedUserEmailAddress).
+		Msg("Creating new user")
+	user := gocloak.User{
+		FirstName: &firstName,
+		LastName:  &lastName,
+		Email:     &invitation.InvitedUserEmailAddress,
+		Enabled:   gocloak.BoolP(true),
+		Username:  &invitation.InvitedUserEmailAddress,
+		Attributes: &map[string][]string{
+			idAttr:       {u.String()},
+			userTypeAttr: {userTypeVal},
+		},
+		RequiredActions: &userRequiredActions,
+	}
+
+	id, err := b.client.CreateUser(ctx, token.AccessToken, b.userRealm, user)
+	if err != nil {
+		b.logger.Error().
+			Str(idAttr, u.String()).
+			Str("email", invitation.InvitedUserEmailAddress).
+			Err(err).
+			Msg("Failed to create user")
+		return "", err
+	}
+
+	return id, nil
+}
+
+// CanSendMail returns true because keycloak does allow to send mail.
+func (b Backend) CanSendMail() bool { return true }
+
+// SendMail sends a mail to the user with details on how to reedeem the invitation.
+func (b Backend) SendMail(ctx context.Context, id string) error {
+	token, err := b.getToken(ctx)
+	if err != nil {
+		return err
+	}
+	params := gocloak.ExecuteActionsEmail{
+		UserID:  &id,
+		Actions: &userRequiredActions,
+	}
+	return b.client.ExecuteActionsEmail(ctx, token.AccessToken, b.userRealm, params)
+}
+
+func (b Backend) getToken(ctx context.Context) (*gocloak.JWT, error) {
+	b.logger.Debug().Msg("Logging into keycloak")
+	token, err := b.client.LoginClient(ctx, b.clientID, b.clientSecret, b.clientRealm)
+	if err != nil {
+		b.logger.Error().Err(err).Msg("failed to get token")
+		return nil, fmt.Errorf("failed to get token: %w", err)
+	}
+
+	rRes, err := b.client.RetrospectToken(ctx, token.AccessToken, b.clientID, b.clientSecret, b.clientRealm)
+	if err != nil {
+		b.logger.Error().Err(err).Msg("failed to introspect token")
+		return nil, fmt.Errorf("failed to retrospect token: %w", err)
+	}
+
+	if !*rRes.Active {
+		b.logger.Error().Msg("token not active")
+		return nil, fmt.Errorf("token is not active")
+	}
+
+	return token, nil
+}
+
+// Quick and dirty way to split the last name off from the first name(s), imperfect, because
+// every culture has a different conception of names.
+func splitDisplayName(displayName string) (string, string) {
```

Comment:
this wont work for spanish names who have multiple last names ...

we may have to either distinguish the name in the api or only send the displayname.

Reply:
This won't work for many names, see also line 141 :)

If I had my way, almost every app would just have a single string for "name".